Fix timespan wrap-around when multiplying back by inverse scale

thomasgassmann opened the pull request:
When multiplying back by the inverse scale of for example TotalMinutes,
a wrap-around is caused, because the nearest representable double is
higher than long.MaxValue. It therefore passes the overflow check and
when the value is casted back to long, it makes the tick count
long.MinValue.

As recommended by @tannergooding, this change uses >= instead of > and
<= instead of <, but does not handle the equal case.

Fixes #31361

As mentioned above, it's still up for discussion whether to handle the `==` explicitely.

thomasgassmann committed:
Fix timespan wrap-around when multiplying back by inverse scale

When multiplying back by the inverse scale of for example TotalMinutes,
a wrap-around is caused, because the nearest representable double is
higher than long.MaxValue. It therefore passes the overflow check and
when the value is casted back to long, it makes the tick count
long.MinValue.

As recommended by @tannergooding, this change uses >= instead of > and
<= instead of <, but does not handle the equal case.

Fix dotnet/corefx#42280

danmosemsft commented:
Thanks for the contribution! Is there a test case to add that would fail before and with this fix no longer fails? My impression is that there would be such a case.

tannergooding commented:
Did we come to a decision on if it was better to throw for `>=` or if it was better to specially handle the `==` case? I had listed both possibilities on my original summary: <https://github.com/dotnet/corefx/issues/42280#issuecomment-548840360>

The drawback to using `>=` is that some values (such as `TimeSpan.MaxValue`) won't roundtrip.

The drawback to using `>` and special casing `==` is that 1023 additional values (`long.MaxValue` through `long.MaxValue + 1023`, inclusive) will also be treated as `TimeSpan.MaxValue`.

Given that we track 100ns ticks, this is no more than ~1/10th of a millisecond inaccuracy in an edge case.

thomasgassmann commented:
@danmosemsft I'm not sure, if I'm missing something, but I think all test cases failling before this fix would also fail now, since I have only extended the range of values for `ticks` in which an exception is thrown.

@tannergooding There hasn't been a decision yet. So far, I have just gone with the `>=` option, though I do agree that it would make for a better API, if `TimeSpan.MaxValue` would roundtrip. Which of these options do you prefer?

tannergooding commented:
@danmosemsft, who owns the types? This probably deserves a small discussion to decide the desired behavior.

danmosemsft commented:
@tannergooding per https://github.com/dotnet/runtime/blob/master/docs/area-owners.md, it is @bartonjs and @joperezr

danmosemsft commented:
My thought fwiw is that round trippability of `TimeSpan.MaxValue` is probably worth more to a developer than having `~1/10th of a millisecond inaccuracy in an edge case.`

stephentoub commented:
@bartonjs and @joperezr, what would you like to do here?

bartonjs commented:
Where does TimeSpan.MaxValue fail to roundtrip?  I just checked on .NET Framework, and `TimeSpan.FromMinutes(TimeSpan.MaxValue.TotalMinutes))` is already an OverflowException.

I'm not sure I really understand what the state was, what the state is, and what the options are; even after reading both the original issue and the comments here multiple times.

tannergooding commented:
It throws on full framework. It doesn't currently throw on .NET Core due to a change in the algorithm

We can either have it throw or we can fix it up to correctly roundtrip, with the "caveat" that some values that are up to ~1/10th of a millisecond greater than `long.MaxValue` will also roundtrip.
I believe the general consensus on https://github.com/dotnet/runtime/pull/1069#issuecomment-571315443 is that the ~1/10th of a millisecond in an edge case isn't worth worrying about and so we should just allow it to roundtrip.

bartonjs commented:
I agree that once we're already doing floating-point math that having a 0.01ms error bar that rounds back down to MaxValue (and rounds back up to MinValue?) is better than having MaxValue throw.

Then the net change from .NET Framework would be that more values work, they just happen to clamp to MaxValue and MinValue.  Significantly (>0.01ms) exceeding MaxValue or MinValue still throws because you've exceeded our tolerance band... right? :smile:.

tannergooding commented:
> Significantly (>0.01ms) exceeding MaxValue or MinValue still throws because you've exceeded our tolerance band... right

Correct, those values will still throw.

bartonjs commented:
Alrighty, let's go ahead and switch to the "Make MaxValue work" plan.

thomasgassmann committed:
Merge remote-tracking branch 'upstream/master' into 42280-timespan-wrap-around

tannergooding commented:
@thomasgassmann, do you plan on updating the PR to handle the equal case now? (No rush, just wanted to confirm it is something you are willing to do for this PR).

thomasgassmann commented:
Hi @tannergooding, I'm still willing to update the PR. Unfortunately I don't have that much time right now, because of mandatory military service. 
I will try to update the PR next weekend.

thomasgassmann committed:
Allow TimeSpan to correctly roundtrip

When creating a `TimeSpan` by either dividing, multiplying or using one of
the `From...` methods, the ticks are multiplied by an inverse scale.
When converting the double-precision value back to a long, this
currently causes incorrect values.
As discussed in dotnet/runtime#1069, the consensus is to make MaxValue
work and to allow it to roundtrip.

Fix dotnet/corefx#42280

thomasgassmann committed:
Update TimeSpanTests to check for roundtrip

As discussed in dotnet/runtime#1069, the plan is to not throw anymore
and instead to roundtrip with these valus.

Fix dotnet/corefx#42280

thomasgassmann commented:
I updated the PR according to the current consensus of this discussion. I also noticed, that the issue with scaling the ticks with a double and then casting back to long also exists in the multiply and divide operator implementations. 
There are still some more test cases to be added though.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/TimeSpan.cs:
```diff
@@ -200,8 +200,17 @@ private static TimeSpan Interval(double value, double scale)
+        {
+            if ((ticks > long.MaxValue) || (ticks < long.MinValue) || double.IsNaN(ticks))
                 throw new OverflowException(SR.Overflow_TimeSpanTooLong);
+            if (ticks == long.MaxValue)
```
```
ticks = Math.Clamp(ticks, long.MinValue, long.MaxValue);
return new TimeSpan((long)ticks);
```
?

thomasgassmann replied:
I'm not sure if it would work like that, since `Math.Clamp` is inclusive, whereas we need an exclusive comparison. So `Math.Clamp` won't handle the `==` case and the less than and greater than case is already handled by the first condition of this method.

tannergooding replied:
Ah right, because `long.MaxValue` is `9223372036854775808` as a double (so it becomes `-9223372036854775808` as a long`).

You should be able to drop the `ticks == long.MinValue`, since that conversion is already correct at least.

thomasgassmann replied:
Right. I dropped the check for `long.MinValue` in 5573f632a64de8db53443888b1d128d54cba0f15.

stephentoub commented:
@thomasgassmann, looks like there's still some feedback to be addressed?  Thanks.

thomasgassmann committed:
TimeSpan: Remove equality check for `long.MinValue`

thomasgassmann commented:
@stephentoub Sorry for the delayed response. I addressed the feedback in the commit above.

stephentoub commented:
@tannergooding, any more feedback?

tannergooding reviewed (approved):
LGTM. It would be good if @bartonjs and/or @joperezr  could sign off as well and merge, since they are the area owners

bartonjs commented on src/libraries/System.Runtime/tests/System/TimeSpanTests.cs:
```diff
@@ -142,6 +142,16 @@ public static void Add_Invalid()
+        public static void FromMinMaxValue_DoesNotThrow()
+        {
+            var maxTimeSpan = TimeSpan.FromDays(TimeSpan.MaxValue.TotalDays);
+            var minTimeSpan = TimeSpan.FromDays(TimeSpan.MinValue.TotalDays);
```
Our usual stance for what makes something "obvious", and thus permits `var`, is only new/cast/as-cast. (No factories, no "but everybody knows").

While I'll grant it's obvious in this context, let's go ahead and specify the type name.

thomasgassmann replied:
Alright, see 4693fe339cfd68ddab97674d42b8d5d4fc4d37d0.

thomasgassmann committed:
Explicitly denote type instead of using var

joperezr reviewed (approved):
LGTM thanks for adding this @thomasgassmann . It is a behavioral breaking change, but it is making this edge case to be correct now so I believe we want to go ahead and take this.

joperezr commented:
cc: @ericstj are we still tracking somewhere a list with breaking changes from previous versions?

danmosemsft commented:
@PriyaPurkayastha can point you to the right list or label to set.

PriyaPurkayastha commented:
Please create a breaking change issue using the following template:
https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md

It will automatically set the "breaking-change" label on that issue and then it will get onto Genevieve's list for publishing to .NET Core docs. cc @gewarren

gewarren commented:
@tannergooding I think you are on point to create the breaking change issue, is that correct?

tannergooding commented:
Yes, will try to get something up tomorrow
